### PR TITLE
CI: Report performance delta via comment

### LINKED ISSUE #2740: CI: Create performance delta comment via Github Action
The performance delta report is a useful feature that is currently implemented via Hydra, which we want to decommission (#2565).

I’ll play around with https://github.com/marketplace/actions/create-or-update-comment#where-to-find-the-id-of-a-comment.

### PR DESCRIPTION
**Pull request by nomeata**
fixes #2740. See CI.md for an overview of how it works.

### CODE REVIEW

**Comment by nomeata** [comment]
This looks plausible. With this merged, I think turning off hydra becomes a viable option.

**Comment by nomeata** [comment]
foo

File: CI.md [hunk 1]
```diff
@@ -118,6 +118,13 @@ Performance changes are known
 For every PR, the developer is told about performance changes relative to the
 merge point, via an continuously updated comment on the PR.
 
+**Implementation (external):**
+ * Steps in the Github Actoin calculates the correct merge base using
```
Comment (crusso) [inline CI.md:122]:
```suggestion
 * Steps in the Github Action calculates the correct merge base using
```

**Review by crusso — APPROVED** [review]
More of rubberstamp than a review. Perhaps @osa1 should take a look.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -1,5 +1,8 @@
 name: "build"
-on: push
+on:
+  push:
+     branches: [ master ]
```
Comment (osa1) [inline .github/workflows/test.yml:4]: Where do we see the results for master? It seems like the code below is only adding PR comments (`github.event_name == 'pull_request'`) so results for master will be lost?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -24,3 +32,41 @@ jobs:
...
+        path: ./perf-delta.txt
+
+    - name: Find performance comment
+      if: runner.os == 'Linux' && github.event_name == 'pull_request'
+      uses: peter-evans/find-comment@v1
+      id: fc
+      with:
+        issue-number: ${{ github.event.pull_request.number }}
+        comment-author: 'github-actions[bot]'
+        body-includes: '<!-- perf comment -->'
```
Comment (osa1) [inline .github/workflows/test.yml:61]: So this updates an existing perf comment, but where do we create a comment with this line for the first time?

I think adding a new comment might also be useful to see changes with commits.

**Review by osa1 — APPROVED** [review]
LGTM. Added some questions to understand this better.

Should we remove ci-pr.nix now? It seems like `perf-delta.nix` is doing a similar thing. Do we need both?

**Comment by nomeata** [comment]
> Should we remove ci-pr.nix now? It seems like perf-delta.nix is doing a similar thing. Do we need both?

I’d suggest to just pull the plug on hydra, and then remove all hydra files.

**Comment by osa1** [comment]
Even if we keep hydra, is there a need to have two bots adding the same comment?

File: .github/workflows/test.yml (hunk 2 above)
Comment (nomeata) [reply .github/workflows/test.yml:4]: Differential CI needs something to compare changes against, i.e. a base branch, and this is about commenting on PRs, which doesn’t exist for pushes to master, so this stuff only happens for PRs.

You can of course always run `nix-build -A tests.perf` on any commit to get the base data, i.e. the CSV file:
```
~/dfinity/motoko $ cat $(nix-build -A tests.perf)/stats.csv
gas/assetstorage;14442
size/assetstorage;165585
gas/qr;1237055836
size/qr;220465
gas/reversi;1305950
size/reversi;160031
```

We may want to publish the data along with other reports (e.g. flamegraphs), see #2568 for that.

File: .github/workflows/test.yml (hunk 3 above)
Comment (nomeata) [reply .github/workflows/test.yml:61]: This Github action creates a commit if there is none, or updates one if there is one – no change in behavior here (if everything works as intended)

**Comment by nomeata** [comment]
> Even if we keep hydra, is there a need to have two bots adding the same comment?

No, if we keep hydra we would remove the code from `ci-pr.nix`. Do we want to keep hydra?

**Comment by osa1** [comment]
@nomeata let's remove ci-pr.nix then, no need to have two bots add the same comment.

**Comment by nomeata** [comment]
Ok, see https://github.com/dfinity/motoko/pull/2745.

But someone (you?) should still tell the IDX team (@nmattia maybe?) that we no longer need hydra, and they can turn it off. And _then_ we can remove all hydra-related files (`ci.nix`, `ci-pr.nix`, maybe others, will check) from the repo.

**Comment by osa1** [comment]
I thought we still need Hydra until flamegraph generation is ported to GitHub CI? https://github.com/dfinity/motoko/issues/2565#issuecomment-900927515

**Comment by nomeata** [comment]
The graphs are still generated, just not published at a easily accessible url. Given that not even you knew about that feature, and it's not used often, it doesn't seem to be that important, so I suggest we don't wait for that.

**Comment by osa1** [comment]
I think there's a problem with this PR. In #2738 I made a change that changes binary sizes and should also affect performance, but the comment (https://github.com/dfinity/motoko/pull/2738#issuecomment-903910303) is not updated.

**Comment by nomeata** [comment]
Too bad we didn’t keep the other bot for a while to compare…

**Comment by osa1** [comment]
I will run the nix command locally to compare, but the comment is updated 16 hours ago so definitely not updated after my last commit which I pushed about an hour ago.

**Comment by nomeata** [comment]
oh, parallel conversation threads. Let’s continue at https://github.com/dfinity/motoko/pull/2738#issuecomment-904432467. I think all is well.

**Comment by nmattia** [comment]
I've pinged IDX for you!

**Pull request merged**